lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -50,9 +51,20 @@ sub _allocate_jobs ($self, $free_workers) {
     my $max_allocate = $limit >= 0 ? min(MAX_JOB_ALLOCATION, $limit - $running) : MAX_JOB_ALLOCATION;
 
     # update the matching workers to the current free
-    for my $jobinfo (values %$scheduled_jobs) {
-        $jobinfo->{matching_workers} = _matching_workers($jobinfo, $free_workers);
+    my %rejected;
+    for my $id (keys %$scheduled_jobs) {
+        my $jobinfo = $scheduled_jobs->{$id};
+        $jobinfo->{matching_workers} = _matching_workers($jobinfo, $free_workers, \%rejected);
+        delete $scheduled_jobs->{$id} unless @{$jobinfo->{matching_workers}};
     }
+    if (keys %rejected) {
+        my @rejected = sort { $rejected{$b} <=> $rejected{$a} || $a cmp $b } keys %rejected;
+        @rejected = @rejected[0 .. MAX_REPORT_MISSING_WORKER_CLASSES-1] if @rejected > MAX_REPORT_MISSING_WORKER_CLASSES;
```

Comment:
This would be simper (no `-1`, no assignment, not `..` syntax) and probably even more performant:
```suggestion
        splice @rejected, MAX_REPORT_MISSING_WORKER_CLASSES if @rejected > MAX_REPORT_MISSING_WORKER_CLASSES;
```

Or even just:
```suggestion
        splice @rejected, MAX_REPORT_MISSING_WORKER_CLASSES;
```